libraries/rush-sdk/src/proxy.ts

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import type { SpawnSyncReturns } from 'child_process';
+import { JsonFile, JsonObject, Executable } from '@rushstack/node-core-library';
+
+import {
+  tryFindRushJsonLocation,
+  RUSH_LIB_NAME,
+  RushLibModuleType,
+  requireRushLibUnderFolderPath,
+  sdkContext
+} from './helpers';
+
+declare const global: NodeJS.Global &
+  typeof globalThis & {
+    ___rush___rushLibModule?: RushLibModuleType;
+    ___rush___rushLibModuleFromEnvironment?: RushLibModuleType;
+    ___rush___rushLibModuleFromInstallAndRunRush?: RushLibModuleType;
+  };
+
+/**
+ * Type of {@link ISdkCallbackEvent.logMessage}
+ * @public
+ */
+export interface IProgressBarCallbackLogMessage {
+  /**
+   * A status message to print in the log window, or `undefined` if there are
+   * no further messages.  This string may contain newlines.
+   */
+  message: string;
+
+  /**
+   * The type of message.  More message types may be added in the future.
+   */
+  messageType: 'info' | 'debug';
+}
+
+/**
+ * Type of {@link ISdkCallbackEvent.state}
+ * @public
+ */
+export type SdkCallbackState =
+  /**
+   * The task has started and is still running.
+   */
+  | 'running'
+  /**
+   * The task has completed successfully. No further events will be fired.
+   */
+  | 'succeeded'
+  /**
+   * The task was aborted by the caller. No further events will be fired.
+   */
+  | 'aborted'
+  /**
+   * The task has failed due to an error. No further events will be fired.
+   */
+  | 'failed';
+
+/**
+ * Event options for {@link ILoadSdkAsyncOptions.onNotifyEvent}
+ * @public
+ */
+export interface ISdkCallbackEvent {
+  /**
+   * Indicates the current state of the operation.
+   */
+  state: SdkCallbackState;
+
+  /**
+   * Allows the caller to display log information about the operation.
+   */
+  logMessage: IProgressBarCallbackLogMessage | undefined;
+
+  /**
+   * Allows the caller to display a progress bar for long-running operations.
+   *
+   * @remarks
+   * If a long-running operation is required, then `progressBarPercent` will
+   * start at 0.0 and count upwards and finish at 100.0 if the operation completes
+   * successfully.  If the long-running operation has not yet started, or
+   * is not required, then the value will be `undefined`.
+   */
+  progressBarPercent: number | undefined;
```

Comment:
Since we have a Promise-based API, the only time we need to report anything outside of the promise channel is for progress events. Success/failure/abort are all within the domain of the promise itself.